pkg/apis/rollouts/validation/validation_references_test.go

```diff
@@ -0,0 +1,271 @@
+	- destination:
+		host: 'not-stable'
+	  weight: 100
+	- destination:
+	    host: canary
+	  weight: 0`
+
+var analysisTemplate = v1alpha1.AnalysisTemplate{
+	ObjectMeta: metav1.ObjectMeta{
+		Name: "analysis-template-name",
+	},
+	Spec: v1alpha1.AnalysisTemplateSpec{
+		Metrics: []v1alpha1.Metric{{
+			Name:     "metric-name",
+			Interval: "1m",
+			Count:    1,
+		}},
+	},
+}
+
+var analysisTemplateWithType = AnalysisTemplateWithType{
+	AnalysisTemplate: &analysisTemplate,
+	TemplateType:     CanaryStep,
+	AnalysisIndex:    0,
+	CanaryStepIndex:  0,
+}
+
+var ro = &v1alpha1.Rollout{
+	Spec: v1alpha1.RolloutSpec{
+		Strategy: v1alpha1.RolloutStrategy{
+			Canary: &v1alpha1.CanaryStrategy{
+				StableService: "stable-service-name",
+				TrafficRouting: &v1alpha1.RolloutTrafficRouting{
+					ALB: &v1alpha1.ALBTrafficRouting{
+						Ingress: "alb-ingress",
+					},
+				},
+			},
+		},
+	},
+}
+
+var ingress = v1beta1.Ingress{
+	ObjectMeta: metav1.ObjectMeta{
+		Name: "alb-ingress",
+	},
+	Spec: v1beta1.IngressSpec{
+		Rules: []v1beta1.IngressRule{{
+			Host: "fakehost.example.com",
+			IngressRuleValue: v1beta1.IngressRuleValue{
+				HTTP: &v1beta1.HTTPIngressRuleValue{
+					Paths: []v1beta1.HTTPIngressPath{{
+						Path: "/foo",
+						Backend: v1beta1.IngressBackend{
+							ServiceName: "stable-service-name",
+							ServicePort: intstr.FromString("use-annotations"),
+						},
+					}},
+				},
+			},
+		}},
+	},
+}
+
+var svc = ServiceWithType{
+	Service: &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "stable-service-name",
+		},
+	},
+	Type: StableService,
+}
+
+func strToUnstructured(yamlStr string) *unstructured.Unstructured {
+	obj := make(map[string]interface{})
+	yamlStr = strings.ReplaceAll(yamlStr, "\t", "    ")
+	err := yaml.Unmarshal([]byte(yamlStr), &obj)
+	if err != nil {
+		panic(err)
+	}
+	return &unstructured.Unstructured{Object: obj}
+}
+
+func TestValidateRolloutReferencedResources(t *testing.T) {
+	refResources := ReferencedResources{
+		AnalysisTemplateWithType: []AnalysisTemplateWithType{analysisTemplateWithType},
+		Ingresses:                []v1beta1.Ingress{ingress},
+		ServiceWithType:          []ServiceWithType{svc},
+		VirtualServices:          nil,
+	}
+	allErrs := ValidateRolloutReferencedResources(ro, refResources)
+	assert.Empty(t, allErrs)
+}
+
+func TestValidateAnalysisTemplateWithType(t *testing.T) {
+	// Success - specify count so that metric runs deterministically
+	allErrs := ValidateAnalysisTemplateWithType(analysisTemplateWithType)
+	assert.Empty(t, allErrs)
+
+	// Fail case - AnalysisTemplate runs indefinitely
+	analysisTemplate.Spec.Metrics[0].Count = 0
+	allErrs = ValidateAnalysisTemplateWithType(analysisTemplateWithType)
+	assert.Len(t, allErrs, 1)
+	assert.Equal(t, "spec.strategy.canary.steps[0].analysis.templates[0].templateName: Invalid value: \"analysis-template-name\": AnalysisTemplate analysis-template-name has metric metric-name which runs indefinitely", allErrs[0].Error())
```

Comment:
This may be easier to read if you create an field.Error struct with all the fields filled out and then do an err.Error() to get the string.